test/apps/iodemo/io_demo.cc

```diff
@@ -134,76 +140,160 @@ const unsigned IoDemoRandom::_M = 0x7fffffffU;
 
     typedef enum {
         XFER_TYPE_SEND,
         XFER_TYPE_RECV
     } xfer_type_t;
-    
+
+    class ChkIoMessage;
+
     /* Asynchronous IO message */
     class IoMessage : public UcxCallback {
+        friend class ChkIoMessage;
     public:
-        IoMessage(size_t buffer_size, MemoryPool<IoMessage>* pool) {
-            _buffer      = malloc(buffer_size);
+        IoMessage(size_t io_msg_size, MemoryPool<IoMessage>* pool) {
+            _buffer      = malloc(io_msg_size);
             _pool        = pool;
-            _buffer_size = buffer_size;
+            _chk_pool    = NULL;
+            _io_msg_size = std::max(io_msg_size, sizeof(iomsg_t));
+            _hdr_size    = 0;
         }
-        
-        void init(io_op_t op, uint32_t sn, size_t data_size) {
-            iomsg_hdr_t *hdr = reinterpret_cast<iomsg_hdr_t*>(_buffer);
-            assert(sizeof(*hdr) <= _buffer_size);
-            hdr->op          = op;
-            hdr->sn          = sn;
-            hdr->data_size   = data_size;
+
+    protected:
+        IoMessage(size_t io_msg_size, MemoryPool<ChkIoMessage>* pool) {
+            _io_msg_size = std::max(io_msg_size,
+                                    sizeof(chk_hdr_t) + sizeof(iomsg_t));
+            _buffer      = malloc(_io_msg_size);
+            _pool        = NULL;
+            _chk_pool    = pool;
+            _hdr_size    = sizeof(chk_hdr_t);
+        }
+
+        iomsg_t* msg() const{
+            return reinterpret_cast<iomsg_t*>((uint8_t*)_buffer + _hdr_size);
         }
 
-        ~IoMessage() {
+    public:
+        virtual void init(io_op_t op, uint32_t sn, size_t data_size) {
+            iomsg_t *m = msg();
+            assert((_hdr_size + sizeof(*m)) <= _io_msg_size);
+            m->tr.sn     = sn;
+            m->op        = op;
+            m->data_size = data_size;
+        }
+
+        virtual ~IoMessage() {
             free(_buffer);
         }
 
         virtual void operator()(ucs_status_t status) {
             _pool->put(this);
         }
 
-        void *buffer() {
+        void *buffer() const {
             return _buffer;
         }
 
+        virtual size_t size() const {
+            assert(sizeof(iomsg_t) <= _io_msg_size);
+            return _io_msg_size;
+        }
+
+        virtual io_op_t op() const {
+            return msg()->op;
+        }
+
+        virtual uint32_t sn() const {
+            return msg()->tr.sn;
+        }
+
     private:
-        void*                  _buffer;
-        MemoryPool<IoMessage>* _pool;
-        size_t                 _buffer_size;
+        void*                     _buffer;
+        MemoryPool<IoMessage>*    _pool;
+        MemoryPool<ChkIoMessage>* _chk_pool;
+        size_t                    _io_msg_size;
+        size_t                    _hdr_size;
+    };
+
+    /* Asynchronous IO message */
+    class ChkIoMessage : public IoMessage {
+    public:
+        ChkIoMessage(size_t buffer_size, MemoryPool<ChkIoMessage>* pool)
```

Comment:
can we unite ChkIoMessage and IoMessage? for simplicity

Reply:
if performance is not critical - yes

Reply:
also it adds ugly code for initialization

Reply:
overall, the code of io_demo would be smaller if we unite, and will not affect performance

Reply:
it will add extra several branches and function parameters even if validation is disabled, IMO separate class keeps all the logic separately and doesn't involve it at all if it's disabled.

Reply:
it is more important to keep the test simple

Reply:
if spaghetti-code is simpler then ok will unite

Reply:
> if spaghetti-code is simpler then ok will unite

IMO it should be only 2 branches: 1. when filling the header put either 0 or real crc(and compute checksum) and 2. when getting the message, wither check the crc or not